File: plugins/togetherai-skills/skills/together-dedicated-containers/scripts/queue_client.ts
#!/usr/bin/env -S npx tsx
/**
 * Together AI Dedicated Containers — Queue Client (TypeScript SDK)
 *
 * Submit jobs, poll for results, and manage queue operations.
 *
 * Usage:
 *     npx tsx queue_client.ts
 *
 * Requires:
 *     npm install together-ai
 *     export TOGETHER_API_KEY=your_key
 *     export TOGETHER_DEPLOYMENT_NAME=your-deployment-name
 */

import Together from "together-ai";

const client = new Together();

const DEPLOYMENT = process.env.TOGETHER_DEPLOYMENT_NAME ?? "hello-world";

async function submitAndPoll(
  payload: Record<string, any>,
  priority: number = 1
): Promise<any> {
  const job = await client.beta.jig.queue.submit({
    model: DEPLOYMENT,
    payload,
    priority,
  });
  const requestId = job.requestId;
  if (!requestId) {
    throw new Error("Queue submit response did not include a request id.");
  }
  console.log(`Submitted job: ${requestId}`);

  while (true) {
    const status: any = await client.beta.jig.queue.retrieve({
      request_id: requestId,
      model: DEPLOYMENT,
    });

    let line = `  Status: ${status.status}`;
    if (status.info?.progress !== undefined) {
      line += ` | Progress: ${(status.info.progress * 100).toFixed(0)}%`;
    }
    console.log(line);

    if (status.status === "done") {
      console.log("  Outputs:", JSON.stringify(status.outputs));
      return { status: "done", outputs: status.outputs };
    } else if (status.status === "failed") {
      console.log("  Error:", status.error);
      return { status: "failed", error: status.error };
    } else if (status.status === "canceled") {
      console.log("  Job was canceled");
      return { status: "canceled" };
    }

    await new Promise((r) => setTimeout(r, 2000));
  }
}

async function submitMultiple(
  payloads: Record<string, any>[]
): Promise<string[]> {
  const requestIds: string[] = [];
  for (const payload of payloads) {
    const job = await client.beta.jig.queue.submit({
      model: DEPLOYMENT,
      payload,
    });
    const requestId = job.requestId;
    if (!requestId) {
      throw new Error("Queue submit response did not include a request id.");
    }
    requestIds.push(requestId);
    console.log(`Submitted: ${requestId}`);
  }
  return requestIds;
}

async function checkStatus(requestId: string): Promise<any> {
  const status: any = await client.beta.jig.queue.retrieve({
    request_id: requestId,
    model: DEPLOYMENT,
  });
  console.log(`Job ${requestId}: ${status.status}`);
  if (status.status === "done") {
    console.log("  Outputs:", JSON.stringify(status.outputs));
  }
  return { status: status.status, outputs: status.outputs };
}

async function main() {
  // --- Example 1: Submit and wait for result ---
  console.log("=== Submit and poll ===");
  await submitAndPoll({ name: "Together" });
  console.log();

  // --- Example 2: Submit with priority ---
  console.log("=== Priority job ===");
  await submitAndPoll({ name: "Priority User" }, 10);
  console.log();

  // --- Example 3: Submit batch ---
  console.log("=== Batch submit ===");
  const ids = await submitMultiple([
    { name: "Alice" },
    { name: "Bob" },
    { name: "Charlie" },
  ]);
  console.log(`Submitted ${ids.length} jobs`);

  for (const rid of ids) {
    await new Promise((r) => setTimeout(r, 1000));
    await checkStatus(rid);
  }
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: plugins/togetherai-skills/skills/together-dedicated-containers/scripts/queue_client.py
#!/usr/bin/env python3
"""
Together AI Dedicated Containers — Queue Client (v2 SDK)

Submit jobs, poll for results, and manage queue operations.

Usage:
    python queue_client.py

Requires:
    uv pip install "together>=2.0.0"
    export TOGETHER_API_KEY=your_key
    export TOGETHER_DEPLOYMENT_NAME=your-deployment-name
"""

import os
import time
from together import Together

client = Together()

DEPLOYMENT = os.environ.get("TOGETHER_DEPLOYMENT_NAME", "hello-world")


def submit_and_poll(payload: dict, priority: int = 1) -> dict:
    """Submit a job and poll until completion."""
    job = client.beta.jig.queue.submit(
        model=DEPLOYMENT,
        payload=payload,
        priority=priority,
    )
    print(f"Submitted job: {job.request_id}")

    while True:
        status = client.beta.jig.queue.retrieve(
            request_id=job.request_id,
            model=DEPLOYMENT,
        )
        print(f"  Status: {status.status}", end="")

        # Show progress if available
        if hasattr(status, "info") and status.info:
            progress = status.info.get("progress")
            if progress is not None:
                print(f" | Progress: {progress:.0%}", end="")
        print()

        if status.status == "done":
            print(f"  Outputs: {status.outputs}")
            return {"status": "done", "outputs": status.outputs}
        elif status.status == "failed":
            print(f"  Error: {getattr(status, 'error', 'unknown')}")
            return {"status": "failed", "error": getattr(status, "error", None)}
        elif status.status == "canceled":
            print("  Job was canceled")
            return {"status": "canceled"}

        time.sleep(2)


def submit_multiple(payloads: list[dict]) -> list[str]:
    """Submit multiple jobs and return their request IDs."""
    request_ids = []
    for payload in payloads:
        job = client.beta.jig.queue.submit(
            model=DEPLOYMENT,
            payload=payload,
        )
        request_ids.append(job.request_id)
        print(f"Submitted: {job.request_id}")
    return request_ids


def check_status(request_id: str) -> dict:
    """Check the status of a single job."""
    status = client.beta.jig.queue.retrieve(
        request_id=request_id,
        model=DEPLOYMENT,
    )
    print(f"Job {request_id}: {status.status}")
    if status.status == "done":
        print(f"  Outputs: {status.outputs}")
    return {"status": status.status, "outputs": getattr(status, "outputs", None)}


if __name__ == "__main__":
    # --- Example 1: Submit and wait for result ---
    print("=== Submit and poll ===")
    result = submit_and_poll({"name": "Together"})
    print()

    # --- Example 2: Submit with priority ---
    print("=== Priority job ===")
    result = submit_and_poll({"name": "Priority User"}, priority=10)
    print()

    # --- Example 3: Submit batch ---
    print("=== Batch submit ===")
    ids = submit_multiple([
        {"name": "Alice"},
        {"name": "Bob"},
        {"name": "Charlie"},
    ])
    print(f"Submitted {len(ids)} jobs")

    # Poll all
    for rid in ids:
        time.sleep(1)
        check_status(rid)
